bcda/models/postgres/alr.go

```diff
@@ -0,0 +1,252 @@
+package postgres
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/CMSgov/bcda-app/bcda/models"
+	"github.com/jackc/pgx"
+	"github.com/jackc/pgx/stdlib"
+	"github.com/sirupsen/logrus"
+)
+
+/*******************************************************************************
+	DATA STRUCTURES
+		1. AlrRepository
+		2. alrCopyFromSource
+			- Used to implement the CopyFrom from the pgx package CopyFrom is
+			Go's connection to PostgreSQL's COPY functionality.
+*******************************************************************************/
+
+type AlrRepository struct {
+	*sql.DB
+	*pgx.Conn
+}
+
+type alrCopyFromSource struct {
+	ctx       context.Context
+	metaKey   int
+	timestamp time.Time
+	alrCount  int
+	alrRowLen int
+	alrs      []models.Alr
+}
+
+/*******************************************************************************
+	HELPER FUNCTIONS
+		1. NewAlrRepo - Used to instantiate AlrRepository struct
+		2. gogEncoder
+			- Used for turning map[string]string to []bytes
+		3. gobDecoder
+			- Used for turning []bytes back to map
+*******************************************************************************/
+
+func NewAlrRepo(db *sql.DB) *AlrRepository {
+	conn, err := stdlib.AcquireConn(db)
+	if err != nil {
+		logrus.Warn(err, "failed to acquire connection")
+	}
+	return &AlrRepository{db, conn}
+}
+
+func GobEncoder(mp map[string]string) ([]byte, error) {
+	var bytea bytes.Buffer
+	enc := gob.NewEncoder(&bytea)
+	if err := enc.Encode(mp); err != nil {
+		return nil, err
+	}
+	return bytea.Bytes(), nil
+}
+
+func GobDecoder(byteMap []byte) (map[string]string, error) {
+	var mp map[string]string
+	r := bytes.NewReader(byteMap)
+	dec := gob.NewDecoder(r)
+	if err := dec.Decode(&mp); err != nil {
+		return nil, err
+	}
+	return mp, nil
+}
+
+/*******************************************************************************
+	METHODS
+		1. Next, Err, Values
+			- These methods are need to implement the CopyFrom from pgx.
+		2. AddAlr
+			- Adds new rows to the ALR Tables. A single call of this function
+			can ingest ALR data per ACO.
+		3. GetAlr
+			- Retreive data from database.
+*******************************************************************************/
+
+func (a *alrCopyFromSource) Next() bool {
+	// If we went through all of the rows, this condition should be false
+	return a.alrCount < a.alrRowLen
+}
+
+func (a *alrCopyFromSource) Err() error {
+	// Err() returns nil if "Done" has not closed and context is still valid
+	return a.ctx.Err()
+}
+
+func (a *alrCopyFromSource) Values() ([]interface{}, error) {
+	// Get the row
+	alr := a.alrs[a.alrCount]
+	// Move counter up so we look at the following row next
+	a.alrCount++
+
+	// Convert the map[string]string to slices of bytes
+	bytea, err := GobEncoder(alr.KeyValue)
+	if err != nil {
+		return nil, err
+	}
+
+	row := []interface{}{
+		// Ordering is based off how they are in the database. See migration 10
```

Comment:
Isn't the ordering based on https://github.com/CMSgov/bcda-app/pull/652/files#diff-baccc3a8849f6a481a2a5e300e1f21d77c7fdda004d8e1157d01f7069af3773aR152

Reply:
Might be premature, but we may want to look at using the pgtype representations directly.

Example: https://github.com/CMSgov/bcda-app/blob/master/bcda/cclf/importer.go#L63


Here's the open issue: https://github.com/jackc/pgx/issues/763

Reply:
In regards to ordering: yes. The order in line 152 is based off the order in database. If anyone is wondering if there was a specific reason for that order.